compiler-plugin/src/main/kotlin/arrow/meta/plugins/union/UnionsPlugin.kt

```diff
@@ -0,0 +1,263 @@
+package arrow.meta.plugins.union
+
+import arrow.meta.Meta
+import arrow.meta.Plugin
+import arrow.meta.invoke
+import arrow.meta.phases.codegen.ir.IrUtils
+import org.jetbrains.kotlin.descriptors.ClassDescriptor
+import org.jetbrains.kotlin.descriptors.findClassAcrossModuleDependencies
+import org.jetbrains.kotlin.diagnostics.Diagnostic
+import org.jetbrains.kotlin.diagnostics.Errors
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.declarations.IrProperty
+import org.jetbrains.kotlin.ir.declarations.IrVariable
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.IrReturn
+import org.jetbrains.kotlin.ir.expressions.IrTypeOperatorCall
+import org.jetbrains.kotlin.ir.expressions.impl.IrGetObjectValueImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrReturnImpl
+import org.jetbrains.kotlin.ir.symbols.impl.IrClassSymbolImpl
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.impl.originalKotlinType
+import org.jetbrains.kotlin.ir.util.dump
+import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.psi2ir.findSingleFunction
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.types.typeUtil.asTypeProjection
+
+/**
+ *
+ * The Union Types Plugin allows users to define typed Unions of arbitrary arity eliminating
+ * nesting and wrappers.
+ *
+ * https://en.wikipedia.org/wiki/Union_type
+ *
+ * ```kotlin:diff
+ * - sealed class Error {
+ * -   object NotANumber : Error()
+ * - }
+ * -
+ * - fun parse(s: String): Either<Error, Int> =
+ * -   if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
+ * -   else Either.Left(Error.NotANumber)
+ * + object NotANumber
+ * + typealias `NotANumber | Int` = Union2<NotANumber, Int>
+ * +
+ * + fun parse(s: String): `NotANumber | Int` =
+ * +  if (s.matches(Regex("-?[0-9]+"))) s.toInt()
+ * +  else NotANumber
+ * ```
+ */
+val Meta.unionTypes: Plugin
+  get() =
+    "Union Types" {
+      meta(
+        /**
+         * ```kotlin
+         * A? : Union2<A, B>
+         * B? : Union2<A, B>
+         *
+         * fun proof(): Union2<String, Int> = 1
+         * ```
+         */
+        typeChecker(::UnionTypeChecker),
+        /**
+         * ```kotlin
+         * fun proof(): Union2<String, Int> = 1
+         * val willBeNullButOk: String? = proof() //ok
+         * val willBeInt: Int? = proof() //ok
+         * ```
+         */
+        suppressDiagnostic(Diagnostic::suppressTypeMismatchOnNullableReceivers),
```

Comment:
We need the same pice in the ide.
May you add that or can I do it?

Reply:
Please go ahead, id love others to understand the plugin